orchagent/orchdaemon.cpp

```diff
@@ -513,6 +517,19 @@ void OrchDaemon::start()
                 }
             }
         }
+        if (gIsSwitchShutdown) {
+            static time_t last_log = 0;
```

Comment:
>static time_t last_log = 0; [](start = 12, length = 27)

define it outside while loop?

Reply:
Prefer to keep it inside the scope, where it is used. Gives better readability.

Reply:
In theory, `static` is shared by all instances of this class, so better to remove `static`

---
In reply to: [448016011](https://github.com/Azure/sonic-swss/pull/1336#discussion_r448016011) [](ancestors = 448016011)

Reply:
I believe when used in this context `static` persists this variable to this particular function call, which lets us define it inside the if-check. I'm not 100% sure what would happen if we had another OrchDaemon, but I think it would define its own `static last_log`.  We could investigate further, but I think it might be simpler just to define this variable outside of the loop as Qi suggested.

Reply:
Inside of a plain C function, a `static` variable is scoped to that function only. However, inside a C++ *class method*, a `static` variable is actually scoped to the class, as Qi stated. Therefore, multiple instances of the class will share the same variable, as if it was defined as a static member variable. Will we ever have more than one OrchDaemon? Probably not, but if we want to be safe, it's better not use `static` here.

Reply:
1) A static variable is declared once and the single instance is shared among all callers within its scope.
2) An instance method, having a static variable, is shared among all instances of the class.
   In our case, there is only one instance of orchdaemon. But even if there are multiple instances, they all share a single instance of this static variable.
   There is no diff between declaring it global vs local, when it is used only locally.
3) Same with static method too.

In short, if there is a static variable, it is only a single instance within its scope.

To help prove, I wrote a small code. I pasted the code and the runtime o/p.

If you need any more clarifications, I will be more than happy to dig deeper.

`
#include <stdio.h>
#include <string>

#define PRINT_ADDRESS(m, i) \
    printf("%s:%d %s: addrof = %p val=%d\n", __FUNCTION__, __LINE__, m, &(i), i)

static int k = 1;

void test_fn()
{   
    static int i = 2;
    PRINT_ADDRESS("global function", i);
}   

class x {
    public:
        void test_instance(const char *m)
        {   
            static int i = 3;
            PRINT_ADDRESS((std::string(m) + " instance method").c_str(), i);
        }
        static void test_static()
        {   
            static int i = 4;                
            PRINT_ADDRESS("class method", i);
        }   
};      

int main()
{   
    x xx, yy, zz;
    int j = 5;

    test_fn();   
    xx.test_instance("xx");
    yy.test_instance("yy");
    zz.test_instance("zz");
    x::test_static();
    xx.test_static()

    PRINT_ADDRESS("file static var", k);
    PRINT_ADDRESS("stack var", j);
    return 0;
}   

`
`
localadmin@remanava-dev-1:~/try/c$ ./t
test_fn:14 global function: addrof = 0x564678ba8014 val=2
test_instance:23 xx instance method: addrof = 0x564678ba8018 val=3
test_instance:23 yy instance method: addrof = 0x564678ba8018 val=3
test_instance:23 zz instance method: addrof = 0x564678ba8018 val=3
test_static:30 class method: addrof = 0x564678ba801c val=4
test_static:30 class method: addrof = 0x564678ba801c val=4
main:47 file static var: addrof = 0x564678ba8010 val=1
main:49 stack var: addrof = 0x7ffdfa949d84 val=5
localadmin@remanava-dev-1:~/try/c$ 
`

Reply:
If you guys still prefer, pushing the local static to outside as file-local, for **readability**, I will go ahead and change it.

From code perspective alone, this is good. If you are not agreeing, please throw me the challenges, as this is the ***best*** way to learn.

Reply:
No more static.